ocp_resources/machineset.py

```diff
@@ -0,0 +1,225 @@
+
+    def __enter__(self):
+        super().__enter__()
+        if self.wait_ready_state and not self.wait(timeout=self.replicas * TIMEOUT):
+            raise MachineSetNotReadyError(
+                f"Machine-set {self.name} has not reached 'ready' state"
+            )
+        return self
+
+    @property
+    def available_replicas(self):
+        return self.instance.status.availableReplicas
+
+    @property
+    def ready_replicas(self):
+        return self.instance.status.readyReplicas
+
+    @property
+    def desired_replicas(self):
+        return self.instance.status.replicas
+
+    @property
+    def provider_spec_value(self):
+        return self.instance.spec.template.spec.providerSpec.value
+
+    def wait(self, timeout=TIMEOUT, sleep=1):
+        """
+        Wait for machine-set to reach 'ready' state.
+
+        Args:
+            timeout (int): maximum time to wait for the 'ready' state.
+            sleep (int): sleep time between each sample.
+
+        Returns:
+            bool: True if machine-set reached 'ready' state, False otherwise.
+        """
+        try:
+            for ready_replicas in TimeoutSampler(
+                wait_timeout=timeout, sleep=sleep, func=lambda: self.ready_replicas
+            ):
+                if ready_replicas and ready_replicas == self.desired_replicas:
+                    return True
+        except TimeoutExpiredError:
+            LOGGER.error(
+                f"Ready replicas: {self.ready_replicas}, desired replicas: {self.desired_replicas}"
+            )
+            return False
+
+    def scale(self, replicas, wait_timeout=300, sleep=1):
+        """
+        Scale down/up a machine-set.
+
+        Args:
+            replicas (int): num of replicas to scale to.
+            wait_timeout (int): maximum time to wait for scaling the machine-set.
+            sleep (int): sleep time between each sample of the machine-set state.
+
+        Returns:
+            bool: True if scaling the machine set was successful, False otherwise.
+        """
+        LOGGER.info(
+            f"Scale machine-set from {self.desired_replicas} replicas to {replicas} replicas"
+        )
+
+        return self.update(
+            patches={"spec": {"replicas": replicas}},
+            wait_timeout=wait_timeout,
+            sleep=sleep,
+        )
+
+    def update(self, patches, wait_timeout=TIMEOUT, sleep=1):
+        """
+        Update machine-set.
+
+        Args:
+            patches (dict): information to update the machine-set.
+            wait_timeout (int): maximum time to wait for machine-set to be ready after updating.
+            sleep (int): sleep time between each sample of the machine-set state.
+
+        Returns:
+            bool: True if updating the machine-set was successful, False otherwise.
+        """
+        LOGGER.info(f"Update machine-set {self.name} to {patches}")
+        ResourceEditor(patches={self: patches}).update()
```

Comment:
you don't need to call ResourceEditor from inside a resource
check how we do it in Deployment.scale_replicas

Reply:
ack